contracts/foundation/service_provider_test.go

```diff
@@ -0,0 +1,29 @@
+package foundation
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+type test1ServiceProvider struct {
+}
+
+type test2ServiceProvider struct {
+	*UnimplementedServiceProvider
+}
+
+func TestUnimplementedServiceProvider(t *testing.T) {
+	var (
+		sp1 = &test1ServiceProvider{}
+		sp2 = &test2ServiceProvider{}
+	)
+
+	_, ok1 := interface{}(sp1).(ServiceProvider)
```

Comment:
S: I think this's unnecessary, we can optimize like below:
```go
var testRegister = 0

type test2ServiceProvider struct {
	*UnimplementedServiceProvider
}

func (u *test2ServiceProvider) Register(app Application) {
	testRegister++
}

sp2.Register(nil)
assert.Equal(t, 1, testRegister)
```

Reply:
Optimized